ui/pages/confirm-transaction/confirm-transaction.component.js

```diff
@@ -19,188 +27,186 @@ import {
       }
     });
 
-    window.addEventListener('beforeunload', this._beforeUnload);
+    window.addEventListener('beforeunload', _beforeUnload);
 
     if (!totalUnapprovedCount && !sendTo) {
       history.replace(mostRecentOverviewPage);
-      return;
-    }
-
-    if (origin !== 'metamask') {
-      getContractMethodData(data);
-    }
+    } else {
+      if (origin !== ORIGIN_METAMASK) {
+        dispatch(getContractMethodData(data));
+      }
 
-    const txId = transactionId || paramsTransactionId;
-    if (txId) {
-      this.props.setTransactionToConfirm(txId);
+      const txId = transactionId || paramsTransactionId;
+      if (txId) {
+        dispatch(setTransactionToConfirm(txId));
+      }
     }
-  }
 
-  componentWillUnmount() {
-    this._beforeUnload();
-    window.removeEventListener('beforeunload', this._beforeUnload);
-  }
+    return () => {
+      _beforeUnload();
+      window.removeEventListener('beforeunload', _beforeUnload);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
 
-  componentDidUpdate(prevProps) {
-    const {
-      setTransactionToConfirm,
-      transaction: { txData: { txParams: { data } = {}, origin } = {} },
-      clearConfirmTransaction,
-      getContractMethodData,
-      paramsTransactionId,
-      transactionId,
-      history,
-      mostRecentOverviewPage,
-      totalUnapprovedCount,
-      setDefaultHomeActiveTabName,
-    } = this.props;
+  useEffect(() => {
+    const { txData: { txParams: { data } = {}, origin } = {} } = transaction;
 
     if (
       paramsTransactionId &&
       transactionId &&
-      prevProps.paramsTransactionId !== paramsTransactionId
+      prevParamsTransactionId !== paramsTransactionId
     ) {
-      clearConfirmTransaction();
-      setTransactionToConfirm(paramsTransactionId);
-      if (origin !== 'metamask') {
-        getContractMethodData(data);
+      dispatch(clearConfirmTransaction());
+      dispatch(setTransactionToConfirm(paramsTransactionId));
+      if (origin !== ORIGIN_METAMASK) {
+        dispatch(getContractMethodData(data));
       }
-    } else if (
-      prevProps.transactionId &&
-      !transactionId &&
-      !totalUnapprovedCount
-    ) {
-      setDefaultHomeActiveTabName('activity').then(() => {
+    } else if (prevTransactionId && !transactionId && !totalUnapprovedCount) {
+      dispatch(setDefaultHomeActiveTabName('activity')).then(() => {
         history.replace(DEFAULT_ROUTE);
       });
     } else if (
-      prevProps.transactionId &&
+      prevTransactionId &&
       transactionId &&
-      prevProps.transactionId !== transactionId
+      prevTransactionId !== transactionId
     ) {
       history.replace(mostRecentOverviewPage);
     }
+  }, [
+    dispatch,
+    transaction,
+    paramsTransactionId,
+    transactionId,
+    history,
+    mostRecentOverviewPage,
+    prevParamsTransactionId,
+    prevTransactionId,
+    totalUnapprovedCount,
+  ]);
```

Comment:
I recently added this: https://github.com/MetaMask/metamask-extension/blob/develop/ui/hooks/useRouting.js

`navigateToMostRecentOverviewPage` can be useful here.

Reply:
I'm leaning toward leaving the code as is. `navigateToMostRecentOverviewPage` uses `history.push` while we use `history.replace` here. We also use `history.replace(DEFAULT_ROUTE)` which still requires `useHistory` or another custom hook. I find we would use relatively the same amount of code lines to call `useHistory` directly as we through custom hooks